internal/database/postgres.go

```diff
@@ -309,66 +311,125 @@ FROM TempDownsampled;
 }
 
 // TODO: consider returning workstationGroup
-func (conn PostgresConn) LatestData() ([]uplink.GpuStatsUpload, error) {
-	// we pull Uuid twice so we can put one into Stat sample and the other into Info
-	rows, err := conn.db.Query(`SELECT g.Machine, g.Uuid, g.Uuid, g.Name,
-			g.Brand, g.DriverVersion, g.MemoryTotal,
-			s.MemoryUtilisation, s.GpuUtilisation, s.MemoryUsed,
-			s.FanSpeed, s.Temp, s.MemoryTemp, s.GraphicsVoltage,
-			s.PowerDraw, s.GraphicsClock, s.MaxGraphicsClock,
-			s.MemoryClock, s.MaxMemoryClock
+func (conn PostgresConn) LatestData() (broadcast.Workstations, error) {
+	// pull all the machines in, then gpus for those machines
+	// has to all be done in a transaction to avoid race conditions
+	tx, err := conn.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[string][]broadcast.Workstation)
+	machines, err := tx.Query(`SELECT GroupName, Hostname, CPU, Motherboard,
+		Notes, LastSeen
+		FROM Machines`)
+	if err != nil {
+		return nil, errors.Join(err, tx.Rollback())
+	}
+
+	for machines.Next() {
+		var groupName *string
+		var machine broadcast.Workstation
+		var lastSeen time.Time
+
+		err = machines.Scan(&groupName, &machine.Name, &machine.CPU,
+			&machine.Motherboard, &machine.Notes, &lastSeen)
+		if err != nil {
+			return nil, errors.Join(err, tx.Rollback())
+		}
+
+		// coalesce null and empty group names to default
+		if groupName == nil || strings.TrimSpace(*groupName) == "" {
+			fallback := DefaultGroup
+			groupName = &fallback
+		}
+
+		machine.LastSeen = time.Since(lastSeen)
+		machine.Gpus = nil
+
+		groups[*groupName] = append(groups[*groupName], machine)
+	}
+
+	// check for error whilst iterating, continuing if it's "no results"
+	err = machines.Err()
+	if err != nil && !errors.Is(sql.ErrNoRows, err) {
+		return nil, errors.Join(err, tx.Rollback())
+	}
+
+	// attach gpus to all machines
+	// can't be done in the previous loop because we can't be iterating
+	// through two queries at once
+	for group := range groups {
```

Comment:
Bar me misunderstanding this code, this shouldn't have an inner loop that defers to a new SQL query. Rather, a clean, well-formatted and efficient SQL query should be constructed to defer this logic to the SQL Engine.

The SQL query execution planner has an almost certain chance of planning something more efficient than this, with much better performance.

Reply:
Unless we find that our database queries are being too slow, I'd like to treat this as a non-blocking style issue. I think the existing query in `getGpus` is already too long to understand quickly, and really don't want to combine that with joining the `machine` table and pulling out all of its fields too. I personally prefer shorter queries that are more decomposed, because I find them easier to understand, debug and modify.

Reply:
Yeah, I agree that it's fine to have kinda sucky SQL queries for now. We can always change them later, and theirs no reason holding up integration on this.